src/battery/HassIntegration.cpp

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <battery/Controller.h>
+#include <battery/Stats.h>
+#include <battery/HassIntegration.h>
+#include <Configuration.h>
+#include <MqttSettings.h>
+#include <MqttHandleHass.h>
+#include <Utils.h>
+#include <__compiled_constants.h>
+
+namespace Batteries {
+
+HassIntegration::HassIntegration(std::shared_ptr<Stats> spStats)
+    : _spStats(spStats) { }
+
+void HassIntegration::hassLoop()
+{
+    auto const& config = Configuration.get();
+    if (!config.Mqtt.Hass.Enabled) { return; }
+
+    // TODO(schlimmchen): this cannot make sure that transient
+    // connection problems are actually always noticed.
+    if (!MqttSettings.getConnected()) {
+        _publishSensors = true;
+        return;
+    }
+
+    if (!_publishSensors ||
+        !_spStats->getManufacturer().has_value() ||
+        !_spStats->getHassDeviceName().has_value()) { return; }
+
+    publishSensors();
+
+    _publishSensors = false;
+}
+
+void HassIntegration::publishSensors() const
+{
+    publishSensor("Manufacturer",          "mdi:factory",        "manufacturer");
+    publishSensor("Data Age",              "mdi:timer-sand",     "dataAge",       "duration", "measurement", "s");
+
+    if (_spStats->isSoCValid()) {
+        publishSensor("State of Charge (SoC)", "mdi:battery-medium", "stateOfCharge", "battery",  "measurement", "%");
+    }
+
+    if (_spStats->isVoltageValid()) {
+        publishSensor("Voltage", "mdi:battery-charging", "voltage", "voltage", "measurement", "V");
+    }
+
+    if (_spStats->isCurrentValid()) {
+        publishSensor("Current", "mdi:current-dc", "current", "current", "measurement", "A");
+    }
```

Comment:
I don't know why i came up with the idea that we need to check if SoC, voltage or current are valid before we can publish the sensor. Sorry!

I think we can assume that every provider (besides mqtt but mqtt should not have hass anyway) will provide those three values, right?

```suggestion
    publishSensor("State of Charge (SoC)", "mdi:battery-medium", "stateOfCharge", "battery",  "measurement", "%");
    publishSensor("Voltage", "mdi:battery-charging", "voltage", "voltage", "measurement", "V");
    publishSensor("Current", "mdi:current-dc", "current", "current", "measurement", "A");
```